packages/@lwc/engine-core/src/framework/profiler.ts

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { startMeasure, endMeasure, MeasurementPhase } from './performance-timing';
+import { VM } from './vm';
+
+function noop(_opId: number, _phase: number, _cmpName: string, _vm_idx: number) {}
+
+let logOperation = noop;
+
+export enum OperationId {
+    constructor = 0,
+    render = 1,
+    patch = 2,
+    connectedCallback = 3,
+    renderedCallback = 4,
+    disconnectedCallback = 5,
+    errorCallback = 6,
+}
+
+enum Phase {
+    Start = 0,
+    Stop = 1,
+}
+
+const opIdToMeasurementPhaseMappingArray: MeasurementPhase[] = [
+    'constructor',
+    'render',
+    'patch',
+    'connectedCallback',
+    'renderedCallback',
+    'disconnectedCallback',
+    'errorCallback',
+];
+
+let profilerEnabled = false;
+let logMarks = false;
+let bufferLogging = false;
+
+if (process.env.NODE_ENV !== 'production') {
+    profilerEnabled = true;
+    logMarks = true;
+    bufferLogging = false;
+}
+
+const profilerStateCallbacks: ((arg0: boolean) => void)[] = [];
+
+function trackProfilerState(callback: (arg0: boolean) => void) {
+    callback(profilerEnabled);
+    profilerStateCallbacks.push(callback);
+}
+
+function logOperationStart(opId: OperationId, vm: VM) {
+    if (logMarks) {
+        startMeasure(opIdToMeasurementPhaseMappingArray[opId], vm);
+    }
+    if (bufferLogging) {
+        const cmpName = getComponentName(vm);
+        logOperation(opId, Phase.Start, cmpName, vm.idx);
+    }
+}
+
+function logOperationEnd(opId: OperationId, vm: VM) {
+    if (logMarks) {
+        endMeasure(opIdToMeasurementPhaseMappingArray[opId], vm);
+    }
+    if (bufferLogging) {
+        const cmpName = getComponentName(vm);
+        logOperation(opId, Phase.Stop, cmpName, vm.idx);
+    }
+}
+
+function getComponentName(vm: VM) {
+    try {
+        return vm.elm.tagName;
```

Comment:
@caridy what would be a better way to get the name of the component ?

Reply:
I pinged Caridy to give insight

Reply:
when will this throw? if you have a VM, it is always associated to an element, unless something changed recently (ping @pmdartus). we don't keep track of the element anymore in the vm, so your two options are: elm (which is what you have here), or use the vm.component.tagName (which ultimate does the same via elm.tagName).

Reply:
I've removed this and switched it to vm.tagName directly.

Reply:
It shouldn't throw since you should get a valid VM passed into this ?